picohttp_t/picohttp_t.c

```diff
@@ -0,0 +1,330 @@
+/*
+* Author: Christian Huitema
+* Copyright (c) 2017, Private Octopus, Inc.
+* All rights reserved.
+*
+* Permission to use, copy, modify, and distribute this software for any
+* purpose with or without fee is hereby granted, provided that the above
+* copyright notice and this permission notice appear in all copies.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL Private Octopus, Inc. BE LIABLE FOR ANY
+* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#ifdef _WINDOWS
+#include "getopt.h"
+#endif
+#include "picoquic.h"
+#include "util.h"
+#include "picoquictest.h"
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+typedef struct st_picoquic_test_def_t {
+    char const* test_name;
+    int (*test_fn)();
+} picoquic_test_def_t;
+
+typedef enum {
+    test_not_run = 0,
+    test_excluded,
+    test_success,
+    test_failed
+} test_status_t;
+
+static const picoquic_test_def_t test_table[] = {
+    { "http0dot9", http0dot9_test },
+    { "h3zero_integer", h3zero_integer_test },
+    { "qpack_huffman", qpack_huffman_test },
+    { "h3zero_parse_qpack", h3zero_parse_qpack_test },
+    { "h3zero_prepare_qpack", h3zero_prepare_qpack_test },
+    { "h3zero_stream_test", h3zero_stream_test },
+    { "parse_demo_scenario", parse_demo_scenario_test },
+    { "h3zero_server", h3zero_server_test },
+    { "h09_server", h09_server_test },
+    { "generic_server", generic_server_test},
+    { "esni", esni_test }
```

Comment:
ESNI test should really be at the QUIC level. ESNI is about TLS, not HTTP. But I understand that it is tied to the HTTP code for now, so I will just open an issue after this PR is merged. Also, the ESNI specification is not finalized yet, so anything there is provisional.